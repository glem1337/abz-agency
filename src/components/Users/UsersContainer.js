import React from 'react';
import {connect} from 'react-redux';

import Users from './Users';
import usersOperations from '../../Redux/Users/UsersOperations';
import usersSelectors from '../../Redux/Users/UsersSelectors';

class UsersContainer extends React.Component {

    componentDidMount() {
        const {getUsers} = this.props;
        let nextUrl = 'https://frontend-test-assignment-api.abz.agency/api/v1/users?page=1&count=6';
        if (window.innerWidth < 768) {
            nextUrl = 'https://frontend-test-assignment-api.abz.agency/api/v1/users?page=1&count=3';
        }
        getUsers(nextUrl);
    };

    handleNextUsersList = () => {
        const {getUsers, nextUrl} = this.props;
        getUsers(nextUrl);
    };

    render() {
        const {usersList, error, isFetching, nextUrl} = this.props;
        return (
            <Users
                nextUrl={nextUrl}
                usersList={usersList}
                handleNextUsersList={this.handleNextUsersList}
                error={error}
                isFetching={isFetching}
            />
        );
    }
}

const mapStateToProps = state => ({
    usersList: usersSelectors.getUsers(state),
    nextUrl: usersSelectors.getNextUrl(state),
    isFetching: usersSelectors.getIsFetching(state),
    error: usersSelectors.getError(state),
});

const mapDispatchToProps = {
    getUsers: usersOperations.getUsers,
};

export default connect(mapStateToProps, mapDispatchToProps)(UsersContainer);
