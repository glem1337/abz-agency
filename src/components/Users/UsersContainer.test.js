import React from 'react';
import {shallow} from 'enzyme'
import configureStore from 'redux-mock-store'
import {UsersContainer} from "./UsersContainer";
import {Provider} from 'react-redux';

// TODO написать тест при нажатии на Load More и протестировать reducer

describe('Users container', () => {

    const initialState = {
        usersList: [],
        nextUrl: '',
        isFetching: false,
        error: null
    };

    const props = {
        usersList: [],
        nextUrl: '',
        isFetching: false,
        error: null,
        getUsers: () => {}
    };

    describe('Users container dispatch `getUsers()` on click Load More', () => {
        let usersContainer, mockFunc;
        beforeEach(() => {
            mockFunc = jest.fn();
            const nextProps = {
                ...props,
                getUsers: mockFunc,
                nextUrl: 'https://site.com',
            };
            usersContainer = shallow(<UsersContainer {...nextProps} />);
        });

        it('renders properly', () => {
            expect(usersContainer).toMatchSnapshot();
        });

        it('dispatches the `getUsers()` method on click', () => {
            usersContainer.find('.users__button').simulate('click');
            expect(mockFunc).toHaveBeenCalledTimes(2);
        });

        afterEach(() => {
            usersContainer.unmount();
            mockFunc = null;
        });

    });


    describe('Users container state', () => {
        const mockStore = configureStore();
        const store = mockStore(initialState);
        const wrapper = shallow(<Provider store={store}><UsersContainer {...props} /></Provider>);
        const usersContainer = wrapper.find('UsersContainer');

        it('renders properly', () => {
            expect(wrapper).toMatchSnapshot();
        });

        it('initialize Users container with initial state', () => {
            expect(usersContainer.props().usersList).toEqual(initialState.usersList);
            expect(usersContainer.props().nextUrl).toEqual(initialState.nextUrl);
            expect(usersContainer.props().isFetching).toEqual(initialState.isFetching);
            expect(usersContainer.props().error).toEqual(initialState.error);
        });

    });

    describe('Users container initial', () => {
        const mockFunc = jest.fn();
        const nextProps = {
            ...props,
            getUsers: mockFunc,
        };
        const usersContainer = shallow(<UsersContainer {...nextProps} />);

        it('renders properly', () => {
            expect(usersContainer).toMatchSnapshot();
        });

        it('dispatches the `getUsers()` method in `componentDidMount()`', () => {
            expect(mockFunc).toHaveBeenCalledTimes(1);
        });

    });

    describe('Users container isFetching', () => {
        const nextProps = {
            ...props,
            isFetching: true,
        };

        const usersContainer = shallow(<UsersContainer {...nextProps} />)

        it('renders preloader', () => {
            expect(usersContainer.find('Loading')).toHaveLength(1);
        });

        it('renders properly', () => {
            expect(usersContainer).toMatchSnapshot();
        });

    });

    describe('Users container with UsersList', () => {
        const nextProps = {
            ...props,
            usersList: [{id: 0, name: 'John Doe'}],
        };

        const usersContainer = shallow(<UsersContainer {...nextProps} />)

        it('UsersList is rendered', () => {
            expect(usersContainer.find('UsersList')).toHaveLength(1);
        });

        it('renders properly', () => {
            expect(usersContainer).toMatchSnapshot();
        });

    });

    describe('Users container with next url', () => {
        let usersContainer;
        const nextProps = {
            ...props,
            nextUrl: 'https://site.com',
        };

        beforeEach(() => {
            usersContainer = shallow(<UsersContainer {...nextProps} />);
        });

        it('renders properly', () => {
            expect(usersContainer).toMatchSnapshot();
        });

        it('button Load More renders', () => {
            expect(usersContainer.find('.users__button')).toHaveLength(1);
        });

        afterEach(() => {
            usersContainer.unmount();
        });

    });

    describe('Users container with Error message', () => {
        const nextProps = {
            ...props,
            error: 'Some error',
        };

        const usersContainer = shallow(<UsersContainer {...nextProps} />)

        it('renders properly', () => {
            expect(usersContainer).toMatchSnapshot();
        });

        it('error renders', () => {
            expect(usersContainer.find('.users__error')).toHaveLength(1);
        });

    });

});