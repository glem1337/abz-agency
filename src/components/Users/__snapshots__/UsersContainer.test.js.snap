// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users container Users container dispatch \`getUsers()\` on click Load More renders properly 1`] = `ShallowWrapper {}`;

exports[`Users container Users container initial renders properly 1`] = `ShallowWrapper {}`;

exports[`Users container Users container isFetching renders properly 1`] = `ShallowWrapper {}`;

exports[`Users container Users container state renders properly 1`] = `ShallowWrapper {}`;

exports[`Users container Users container with Error message renders properly 1`] = `ShallowWrapper {}`;

exports[`Users container Users container with UsersList renders properly 1`] = `ShallowWrapper {}`;

exports[`Users container Users container with next url renders properly 1`] = `ShallowWrapper {}`;
